clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy

```diff
@@ -177,153 +104,224 @@ class LoadBalancerV2UpsertHandler {
+    } catch (AmazonServiceException e) {
+      String exceptionMessage = "Failed to add listener to ${loadBalancer.loadBalancerName} (${listener.port}:${listener.protocol}) - reason: ${e.errorMessage}."
+      task.updateStatus BASE_PHASE, exceptionMessage
+      amazonErrors << exceptionMessage
+      return false
+    }
 
-      Map<Listener, List<Rule>> removedRules = new HashMap<>()
-      listenersToRemove.each {
-        // Remove rules first
-        List<Rule> rulesToRemove = loadBalancing.describeRules(new DescribeRulesRequest().withListenerArn(it.listenerArn))?.rules
-        // We cannot remove the default rule(s)
-        rulesToRemove.removeAll { it.isDefault }
-        removedRules.put(it, rulesToRemove)
-        rulesToRemove.each { rule ->
-          loadBalancing.deleteRule(new DeleteRuleRequest().withRuleArn(rule.ruleArn))
-          task.updateStatus BASE_PHASE, "Rule associated with listener removed (${rule.ruleArn})."
+    if (result != null && result.listeners.size() > 0) {
+      String listenerArn = result.listeners.get(0).listenerArn
+      try {
+        rules.each { rule ->
+          loadBalancing.createRule(new CreateRuleRequest(listenerArn: listenerArn, conditions: rule.conditions, actions: rule.actions, priority: Integer.valueOf(rule.priority)))
         }
+      } catch (AmazonServiceException e) {
+        String exceptionMessage = "Failed to add rule to listener ${loadBalancer.loadBalancerName} (${listener.port}:${listener.protocol}) reason: ${e.errorMessage}."
+        task.updateStatus BASE_PHASE, exceptionMessage
+        amazonErrors << exceptionMessage
+        return false
+      }
+    }
 
+    return true
+  }
+
+  static void removeListeners(List<Listener> listenersToRemove, List<Listener> existingListeners, AmazonElasticLoadBalancing loadBalancing, LoadBalancer loadBalancer) {
+    listenersToRemove.each {
+      try {
         loadBalancing.deleteListener(new DeleteListenerRequest().withListenerArn(it.listenerArn))
-        task.updateStatus BASE_PHASE, "Listener removed from ${loadBalancerName} (${it.port}:${it.protocol})."
+        task.updateStatus BASE_PHASE, "Listener removed from ${loadBalancer.loadBalancerName} (${it.port}:${it.protocol})."
         existingListeners.remove(it)
+      } catch (ListenerNotFoundException e) {
+        task.updateStatus BASE_PHASE, "Failed to delete listener ${it.listenerArn}. Listener could not be found. ${e.errorMessage}"
       }
+    }
+  }
+
+  static void updateLoadBalancer(AmazonElasticLoadBalancing loadBalancing, LoadBalancer loadBalancer,
+                                        Collection<String> securityGroups,
+                                        List<UpsertAmazonLoadBalancerV2Description.TargetGroup> targetGroups,
+                                        List<UpsertAmazonLoadBalancerV2Description.Listener> listeners) {
+    def amazonErrors = []
```

Comment:
I noticed that the operation keeps on going and just accumulates error messages as it goes.

I wonder if there are some points it should bail out if it has encountered failures?  Perhaps not deleting anything if the creation phases failed?

Reply:
I do check if listeners were all created on line 318 before removing the old ones, but good call, I'll check if there's more places to safeguard.